src/derive_utils.rs

```diff
@@ -10,106 +10,245 @@ use crate::exceptions::PyTypeError;
+        }
+
+        // Handle keyword arguments
+        let varkeywords = match (kwargs, self.accept_varkeywords) {
+            (Some(kwargs), true) => {
+                let mut varkeywords = None;
+                self.extract_keyword_arguments(kwargs, output, |name, value| {
+                    varkeywords
+                        .get_or_insert_with(|| PyDict::new(kwargs.py()))
+                        .set_item(name, value)
+                })?;
+                varkeywords
+            }
+            (Some(kwargs), false) => {
+                self.extract_keyword_arguments(kwargs, output, |name, _| {
+                    Err(self.unexpected_keyword_argument(name))
+                })?;
+                None
+            }
+            (None, _) => None,
+        };
+
+        // Check that there's sufficient positional arguments once keyword arguments are specified
+        if args_provided < self.required_positional_parameters {
+            let missing_positional_arguments: Vec<_> = self.positional_parameter_names
+                [..self.required_positional_parameters]
+                .iter()
+                .copied()
+                .zip(output.iter())
+                .filter_map(|(param, out)| if out.is_none() { Some(param) } else { None })
+                .collect();
+            if !missing_positional_arguments.is_empty() {
+                return Err(
+                    self.missing_required_arguments("positional", &missing_positional_arguments)
+                );
+            }
+        }
+
+        // Check no missing required keyword arguments
+        let missing_keyword_only_arguments: Vec<_> = self
+            .keyword_only_parameters
+            .iter()
+            .zip(&output[num_positional_parameters..])
+            .filter_map(|(keyword_desc, out)| {
+                if keyword_desc.required && out.is_none() {
+                    Some(keyword_desc.name)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        if !missing_keyword_only_arguments.is_empty() {
+            return Err(self.missing_required_arguments("keyword", &missing_keyword_only_arguments));
+        }
+
+        Ok((varargs, varkeywords))
     }
-    // Copy kwargs not to modify it
-    let kwargs = match kwargs {
-        Some(k) => Some(k.copy()?),
-        None => None,
-    };
-    // Iterate through the parameters and assign values to output:
-    for (i, (p, out)) in params.iter().zip(output).enumerate() {
-        *out = match kwargs.and_then(|d| d.get_item(p.name)) {
-            Some(kwarg) => {
-                if i < nargs {
-                    raise_error!("got multiple values for argument: {}", p.name)
+
+    #[inline]
+    fn extract_keyword_arguments<'p>(
+        &self,
+        kwargs: &'p PyDict,
+        output: &mut [Option<&'p PyAny>],
+        mut unexpected_keyword_handler: impl FnMut(&'p PyAny, &'p PyAny) -> PyResult<()>,
```

Comment:
How about returning `Vec<(&'p PyAny, &'p PyAny)>` instead of taking this closure?
Then we can display all unexpected keyword args.

Reply:
I want to avoid `Vec` in the case where we accept `**kwargs`, because it's an unnecessary allocation. However I agree we can display all unexpected keyword arguments - I'll change the code to do so.

Reply:
Actually I've changed my mind on changing what I've currently got - the existing code for printing all unexpected positional arguments relies on having `&str` forms of the parameter names, whereas here we only have `&PyAny` which are not guaranteed to be UTF8.

Also the current form I've implemented is consistent with CPython which only lists the first unexpected keyword argument.

If someone badly wants the full list printed, it can be a follow-up PR 😄

Reply:
> Also the current form I've implemented is consistent with CPython which only lists the first unexpected keyword argument.

OK, that makes sense 👍🏼